lib/parser/builders/default.rb

```diff
@@ -817,6 +812,54 @@ def objc_restarg(star_t, name=nil)
       end
     end
 
+    # 2.7 numbered parameters
+
+    # Given an s(:ident, :var_name) node checks and registers if needed
+    # a numbered parameter in a current block scope
+    def check_and_register_numparam(ident_node)
+      name = ident_node.children[0]
+      location = ident_node.loc.expression
+
+      if name =~ /\A_[1-9]\z/ && !@parser.static_env.declared?(name)
+        # definitely an implicit param
+
+        if !@parser.context.in_block? && !@parser.context.in_lambda?
+          diagnostic :error, :numparam_outside_block, nil, location
+        end
+
+        if @parser.max_numparam_stack.has_ordinary_params?
+          diagnostic :error, :ordinary_param_defined, nil, location
+        end
+
+        raw_context = @parser.context.stack.dup
+        raw_max_numparam_stack = @parser.max_numparam_stack.stack.dup
+
+        # ignore current block scope
+        raw_context.pop
+        raw_max_numparam_stack.pop
+
+        raw_context.reverse_each do |outer_scope|
+          if outer_scope == :block || outer_scope == :lambda
+            outer_scope_has_numparams = raw_max_numparam_stack.pop > 0
+
+            if outer_scope_has_numparams
+              diagnostic :error, :numparam_used_in_outer_scope, nil, location
+            else
+              # for now it's ok, but an outer scope can also be a block
+              # with numparams, so we need to continue
+            end
+          else
+            # found an outer scope that can't have numparams
+            # like def/class/etc
+            break
+          end
+        end
+
+        @parser.static_env.declare(name)
+        @parser.max_numparam_stack.register(name[1].to_i)
+      end
+    end
```

Comment:
My gut feel is that this shouldn't be in the builder. But maybe we don't have the infrastructure to do it any other way.

Reply:
Because it doesn't look generic and may change in the next version of MRI? I guess you are right. Would you move it straight into parser?

Reply:
Yeah. I'm not really sure what's best here. But moving it into parser is one option.

There are other places where this happens, like `accessible` and `assignable`, but those seem more straightforward (and we can't change them anyway due to backwards compat).

Reply:
ok, I moved it straight into parser. I tried implementing it in a more declarative way using enumerators and objects with predicates around stack values, but it still looks complicated. I feel like it should be extracted into a separate class, but then comes the problem of changing it for next rubies. I guess inlining is the best solution.